modules/files.nix

```diff
@@ -197,6 +197,37 @@ in
         ''
     );
 
+    home.activation.checkFilesChanged = dag.entryBefore ["linkGeneration"] (
+      ''
+        declare -A changed_files
+      '' + concatStrings (
+        mapAttrsToList (n: v: ''
+
+        changed_files["${v.target}"]=0
+        cmp --quiet \
+        "${v.source}" \
+        "${v.target}" \
+        || changed_files["${v.target}"]=1
+        '')
+        (filterAttrs (n: v: v.onChange != null)
+        cfg)
+      )
```

Comment:
I don't have a strong opinion but perhaps something along the lines of

```nix
home.activation.checkFilesChanged = dag.entryBefore ["linkGeneration"] (
  ''
    declare -A changed_files
  '' + concatMapStrings (v: ''
    cmp --quiet "${v.source}" "${v.target}" \
      && changed_files["${v.target}"]=0 \
      || changed_files["${v.target}"]=1
  '') (filter (v: v.onChange != "") (attrValues cfg))
)
```

would be a bit cleaner.

The `(filter (v: v.onChange != "") (attrValues cfg))` part also holds for `home.activation.onFilesChange` below.